### PR TITLE
Update and improve ReplicationController resource lifecycle test

### PR DESCRIPTION
**Pull request by BobyMCbobs**
- [x] [APISnoop mock ticket](https://github.com/cncf/apisnoop/pull/289)
- [x] [mock template ticket](#)
- [x] [PR](https://github.com/kubernetes/kubernetes/pull/88588)
- [ ] [update and improve](#)

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Updates the test to rely on responses from `Patch` or `Update` instead of later `Get` or `List`

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/kubernetes/pull/88588#issuecomment-607366847

**Special notes for your reviewer**:
Attempts to fix flakiness.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs
NONE
```

/sig testing

### CODE REVIEW

File: test/e2e/apps/rc.go [hunk 1]
```diff
@@ -260,7 +280,7 @@ var _ = SIGDescribe("ReplicationController", func() {
 
 		ginkgo.By("waiting for ReplicationController to have a DELETED watchEvent")
 		for watchEvent := range rcWatchChan {
-			if watchEvent.Type == "DELETED" {
+			if watchEvent.Type == watch.Deleted {
```
Comment (spiffxp) [inline test/e2e/apps/rc.go:283]: I have the same concern as https://github.com/kubernetes/kubernetes/pull/89707 that this test will not fail if the watch times out

**Comment by spiffxp** [comment]
ref: https://github.com/kubernetes/kubernetes/pull/88588#issuecomment-607366847

> Choices are:
> - patch something the controller isn't messing with regularly (not sure what that would be)

The test covers fewer endpoints if we do this, part of the point of writing it was to patch scale and status, so I would not recommend this approach

> - properly handle optimistic concurrency by retrying in these cases, AND by not necessarily expecting the value we write to be there later

Agree we should do this, and that's not what this PR does.  It's not clear to me that the watches this PR adds are of any benefit.  I would recommend looking around e2e tests to see if there's any prior art you can use

**Comment by spiffxp** [comment]
Maybe take a look at this and how it's used in PDB tests https://github.com/kubernetes/kubernetes/blob/871ca4b2a75decc0a810b804252dff493a20984c/test/e2e/apps/disruption.go#L391-L404

**Comment by spiffxp** [comment]
ref: https://github.com/kubernetes/kubernetes/issues/89740

**Comment by spiffxp** [comment]
pinging for progress here, we're still seeing flakes, if we don't resolve soon I'm inclined to mark the test as `[Flaky]` to get it out of high traffic

**Comment by hh** [comment]
Let's see if this removes the flakes.

File: test/e2e/apps/rc.go [hunk 2]
```diff
@@ -182,9 +210,9 @@ var _ = SIGDescribe("ReplicationController", func() {
 		json.Unmarshal(rcStatusUjson, &rcStatus)
 		framework.ExpectEqual(rcStatus.Status.Replicas, testRcInitialReplicaCount, "ReplicationController ReplicaSet cound does not match initial Replica count")
 
-		rcScalePatchPayload, err := json.Marshal(autoscalingv1.Scale{
-			Spec: autoscalingv1.ScaleSpec{
-				Replicas: testRcMaxReplicaCount,
+		rcScalePatchPayload, err := json.Marshal(v1.ReplicationController{
```
Comment (liggitt) [inline test/e2e/apps/rc.go:213]: this change does not make sense... this is marshaling replication controller to JSON to send to an endpoint that expects autoscaling.Scale JSON... it coincidentally happens to match, but only by coincidence

File: test/e2e/apps/rc.go [hunk 3]
```diff
@@ -139,6 +139,12 @@ var _ = SIGDescribe("ReplicationController", func() {
```
Comment (liggitt) [inline test/e2e/apps/rc.go:139]: using raw watches is fragile; those are subject to being dropped and nothing in this test handles re-establishing the watch

File: test/e2e/apps/rc.go [hunk 4]
```diff
@@ -156,8 +162,15 @@ var _ = SIGDescribe("ReplicationController", func() {
 		framework.ExpectNoError(err, "failed to marshal json of replicationcontroller label patch")
 		// Patch the ReplicationController
 		ginkgo.By("patching ReplicationController")
-		_, err = f.ClientSet.CoreV1().ReplicationControllers(testRcNamespace).Patch(context.TODO(), testRcName, types.StrategicMergePatchType, []byte(rcLabelPatchPayload), metav1.PatchOptions{})
+		testRcPatched, err := f.ClientSet.CoreV1().ReplicationControllers(testRcNamespace).Patch(context.TODO(), testRcName, types.StrategicMergePatchType, []byte(rcLabelPatchPayload), metav1.PatchOptions{})
 		framework.ExpectNoError(err, "Failed to patch ReplicationController")
+		framework.ExpectEqual(testRcPatched.ObjectMeta.Labels["test-rc"], "patched", "failed to patch RC")
+		ginkgo.By("waiting for RC to be modified")
+		for watchEvent := range rcWatchChan {
+			if watchEvent.Type == watch.Modified {
```
Comment (liggitt) [inline test/e2e/apps/rc.go:170]: what is the intent of this loop? other things (like the controller) could produce modified events on an RC besides the patch

File: test/e2e/apps/rc.go [hunk 5]
```diff
@@ -172,6 +185,21 @@ var _ = SIGDescribe("ReplicationController", func() {
 		rcStatus, err := f.ClientSet.CoreV1().ReplicationControllers(testRcNamespace).Patch(context.TODO(), testRcName, types.StrategicMergePatchType, []byte(rcStatusPatchPayload), metav1.PatchOptions{}, "status")
 		framework.ExpectNoError(err, "Failed to patch ReplicationControllerStatus")
 		framework.ExpectEqual(rcStatus.Status.ReadyReplicas, int32(0), "ReplicationControllerStatus's readyReplicas does not equal 0")
+		ginkgo.By("waiting for RC to be modified")
+		for watchEvent := range rcWatchChan {
+			if watchEvent.Type == watch.Modified {
```
Comment (liggitt) [inline test/e2e/apps/rc.go:190]: same question about the intent of this loop... what is it intending to wait for?

**Comment by johnbelamaric** [comment]
We probably should go ahead and disable this due to flakiness until it's fixed. The plan now is to build a utility that will handle the watch restarts, accumulate the events, and have the test process those accumulated events, rather than the raw watch output.

File: test/e2e/apps/rc.go (hunk 2 above)
Comment (BobyMCbobs) [reply test/e2e/apps/rc.go]: Reverted in 9660e480c76

**Comment by BobyMCbobs** [comment]
In the meanwhile, should the watches be updated to have conditions to expect that the events will take place? - like https://github.com/kubernetes/kubernetes/pull/89707#discussion_r412437587

**Comment by johnbelamaric** [comment]
There are now several tests that would need that treatment. I think your time is better spent on getting the helper done and then updating the tests to use those.

**Comment by BobyMCbobs** [comment]
I rebased onto _ii:create-replicationcontroller-retry-tooling_ locally and wired this test up to it:
```go
		rcScalePatchFunc := func(rc *v1.ReplicationController) (bytes []byte, err error) {
			rcScalePatchBytes, err := json.Marshal(rc)
			framework.ExpectNoError(err, "Failed to marshal json of replicationcontroller label patch")
			return rcScalePatchBytes, err
		}

		// Patch the ReplicationController's scale
		ginkgo.By("patching ReplicationController scale")
		_ = patchRCOrDie(f.ClientSet, dc, testRcNamespace, testRcName, rcScalePatchFunc, "scale")
```

At this point, the test fails with
```
E0507 15:59:16.554998   28154 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or ni
l pointer dereference
```

**Comment by hh** [comment]
Closing in favour of: #90879
/close

**Pull request closed without merging**